File: 463.Island_Perimeter.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
    	int cnt=0,count=0,row=grid.size(),cols=grid[0].size();
    	for(int i=0;i<row;i++)
    	{
    		for(int j=0;j<cols;j++)
    		{
    			if(grid[i][j]==1)
    			{
    				cnt++;
				
                    if(j+1<cols and grid[i][j+1]==1)
                        count++;
                    if(i+1<row and grid[i+1][j]==1)
                        count++;
                    if(j-1>=0 and grid[i][j-1]==1)
                        count++;
                    if(i-1>=0 and grid[i-1][j]==1)
                        count++;
			}
			}
		}
		return 4*cnt-count;
        }
};

File: 463.Island_Perimeter.java
class Solution {
    public int islandPerimeter(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        boolean[][] visited = new boolean[m][n];
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j]==1)
                {
                    return dfs(grid, i, j, visited);
                }
            }
        }
        return 0;
    }
    
    private int dfs(int[][] grid, int i, int j, boolean[][] visited)
    {
        if(i<0 || j<0 || i>=grid.length || j>=grid[0].length || grid[i][j] == 0)
            return 1;
        
        if(visited[i][j])
            return 0;
        
        visited[i][j] = true;
        return dfs(grid, i-1, j, visited)+ dfs(grid, i+1, j, visited) + dfs(grid, i, j-1, visited) + dfs(grid, i, j+1, visited);
    }
}
